dashboard/signals.py

```diff
@@ -0,0 +1,33 @@
+"""
+Signals for user profiles
+"""
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from dashboard.models import CachedEnrollment, ProgramEnrollment
+
+
+@receiver(post_save, sender=CachedEnrollment, dispatch_uid="update_programenrollment")
+def create_programenrollment(sender, instance, created, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to create Program enrollment when the CachedEnrollment table is updated
+    """
+    user = instance.user
+    program = instance.course_run.course.program
+    if instance.data is not None:
+        ProgramEnrollment.objects.get_or_create(
+            user=user,
+            program=program
+        )
+    else:
+        # check if there is any other entry with data != None
+        # if not delete the enrollment
+        active_enrollment_count = CachedEnrollment.objects.filter(
+            user=user,
+            course_run__course__program=program
+        ).exclude(data__isnull=True).count()
+        if active_enrollment_count == 0:
+            ProgramEnrollment.objects.filter(
+                user=user,
+                program=program
+            ).delete()
```

Comment:
I think this is a good approach: I like the idea. 
I thought about it and I implemented an even more (I hope) optimized version